06.Lesson_ThinkingProperties/ThinkingPropertiesExercise/MeetingScheduler/README.md

```diff
@@ -0,0 +1,35 @@
+
+- ❓ Is there a way to check that a Meeting always starts as UNITIALIZED?
+- Unit test that confirms that the start and end times provided when creating a meeting are properly saved onto the meeting
+- Check that only the owner can cancel the meeting
+    - ❓ In my test for this (`onlyOwnerCanCancel`), I do a `scheduleMeeting` so that I have a handle on the meeting owner, and then check that `cancelMeeting` can only be called by the owner. But I'm limiting myself to cases where they are called back-to-back, so this test is not as generic as I would like. Would it be better to "harness" a getter for the owner, so that I can have the meeting start in a more arbitrary state.
+    - ❓ My test for this is failing, can you tell why?
+- Meeting can be started by anyone after startTime and beforeEnd time (check this with `lastReverted` etc)
+- A meeting can be joined if it is STARTED (can be joined even if we have passed the endTime, so long as no one has ENDed the meeting)
+    - ❓ My test for this is failing, can you tell why?
+- The owner of a meeting cannot change
+    - Don't think I can test this without access to the `organizer`
+
+# Categories
+
+## Valid states
+- I generally don't understand this category
+- ❓ Is it fair to say that these will usually be proved by `invariant`s rather than `rule`s?
```

Comment:
Yeah, here you can say something like a meeting is only in one state, and define properties of that state. Being as thorough as possible for the state definition is recommended, so thinking of all the side effects of a meeting being uninitialized and proving that they are always true (uninitialized means no participants, start and end time 0, etc.)

Reply:
Can you point me to examples of rules in this category for other contracts?

> here you can say something like a meeting is only in one state

Would this be an invariant like `getStateById(meetingId) == 0 || getStateById(meetingId) == 1 || ...`?  That doesn't seem very interesting. I suspect you are wanting something more like `meetingIsUnitialized(meetingId) || meetingIsPending(meetingId) || ...` where each of the functions like `meetingIsUnitialized` is comprised of a bunch of requirements, as you described at the end.


> so thinking of all the side effects of a meeting being uninitialized and proving that they are always true

What would this sort of proof look like? It all makes sense for "unitialized", but when I think about "pending" it becomes less clear how to prove it. I suppose I could make an invariant about unitialized meetings, and then `requireInvariant` that at the start of my invariant about pending meetings.

Reply:
Have you watched the videos on Aave? If not watch it here: https://www.youtube.com/watch?v=QukwpzHhPL0
Here is a list of properties for Saddle's Swap contract: https://docs.google.com/document/d/12B6ei4-nowkobbSwBUe53RpR6cgscYB5KxAlbmqXMaA/edit

> Would this be an invariant like getStateById(meetingId) == 0 || getStateById(meetingId) == 1 || ...? That doesn't seem very interesting. I suspect you are wanting something more like meetingIsUnitialized(meetingId) || meetingIsPending(meetingId) || ... where each of the functions like meetingIsUnitialized is comprised of a bunch of requirements, as you described at the end.

The second option is what I meant.

> What would this sort of proof look like? It all makes sense for "unitialized", but when I think about "pending" it becomes less clear how to prove it. I suppose I could make an invariant about unitialized meetings, and then requireInvariant that at the start of my invariant about pending meetings.

For pending it would be that the organizer is non-zero, the start and end times are set, the number of participants is 0 etc. All the side effects expected when this change happens.